pandas/core/frame.py

```diff
@@ -8417,6 +8417,132 @@ def isin(self, values):
+        column.
+
+        By default, rows that contain any NaN value are omitted from the
+        results.
+
+        By default, the resulting series will be in descending order so that the
+        first element is the most frequently-occurring row.
+
+        Parameters
+        ----------
+        normalize : boolean, default False
+            If True then the Series returned will contain the relative
+            frequencies of the unique values.
+        sort : boolean, default True
+            Sort by frequencies.
+        ascending : boolean, default False
+            Sort in ascending order.
+        bins : integer, optional
+            Rather than count values, group them into half-open bins,
+            a convenience for ``pd.cut``, only works with single-column numeric
+            data.
+        dropna : boolean, default True
+            Don't include counts of rows containing NaN.
+
+        Returns
+        -------
+        counts : Series
+
+        See Also
+        --------
+        Series.value_counts: Equivalent method on Series.
+
+        Examples
+        --------
+
+        >>> df = pd.DataFrame({'num_legs': [2, 4, 4, 6],
+        ...                    'num_wings': [2, 0, 0, 0]},
+        ...                   index=['falcon', 'dog', 'cat', 'ant'])
+        >>> df
+                num_legs  num_wings
+        falcon         2          2
+        dog            4          0
+        cat            4          0
+        ant            6          0
+
+        >>> df.value_counts()
+        num_legs  num_wings
+        4         0            2
+        6         0            1
+        2         2            1
+        dtype: int64
+
+        >>> df.value_counts(sort=False)
+        num_legs  num_wings
+        2         2            1
+        4         0            2
+        6         0            1
+        dtype: int64
+
+        >>> df.value_counts(ascending=True)
+        num_legs  num_wings
+        2         2            1
+        6         0            1
+        4         0            2
+        dtype: int64
+
+        >>> df.value_counts(normalize=True)
+        num_legs  num_wings
+        4         0            0.50
+        6         0            0.25
+        2         2            0.25
+        dtype: float64
+
+        >>> single_col_df = df[['num_legs']]
+        >>> single_col_df.value_counts(bins=4)
+        num_legs
+        (3.0, 4.0]      2
+        (5.0, 6.0]      1
+        (1.995, 3.0]    1
+        (4.0, 5.0]      0
+        dtype: int64
+        """
+
+        # Some features not supported yet.
+        if not dropna:
+            raise NotImplementedError(
+                "`dropna=False` not yet supported for dataframes."
+            )
+        if bins is not None:
+            raise NotImplementedError(
+                "`bins` parameter not yet supported for dataframes."
+            )
+
+        # Delegate to Series.value_counts for single-column data frames.
```

Comment:
Hmm @TomAugspurger this is the special casing you were referring to right? Sorry kind of tough to tell from history at this point.

Is there a particular reason why we would want to do this? I think the type of output should match input

Reply:
I think the problem was groupby returns a regular Index when you're grouping by a single column? And in all other cases you get a MultiIndex. So we need some kind of special condition for Series to ensure we get a 1-level MI back.

I think it'd be clearer to just do the `counts = self.groupby(self.columns.tolist()).size()` and then follow it up with a

```python
if len(self.columns) == 1:
    counts.index = pd.MultiIndex.from_arrays([counts.index])
```

Reply:
Done. This was leftover from the special casing of parameters for single columns.